types/src/transaction/mod.rs

```diff
@@ -538,6 +762,158 @@ impl TransactionPayload {
             payload => panic!("Expected EntryFunction(_) payload, found: {:#?}", payload),
         }
     }
+
+    pub fn has_nonce(&self) -> bool {
+        match self {
+            Self::Payload(TransactionPayloadInner::V1 {
+                extra_config:
+                    TransactionExtraConfig::V1 {
+                        replay_protection_nonce,
+                        ..
+                    },
+                ..
+            }) => replay_protection_nonce.is_some(),
+            _ => false,
+        }
+    }
+
+    pub fn executable(&self) -> TransactionExecutable {
+        match self {
+            TransactionPayload::EntryFunction(entry_function) => {
+                TransactionExecutable::EntryFunction(entry_function.clone())
+            },
+            TransactionPayload::Script(script) => TransactionExecutable::Script(script.clone()),
+            TransactionPayload::Multisig(multisig) => multisig.as_transaction_executable(),
+            TransactionPayload::Payload(TransactionPayloadInner::V1 { executable, .. }) => {
+                executable.clone()
+            },
+            TransactionPayload::ModuleBundle(_) => {
+                unimplemented!("ModuleBundle variant is deprecated")
+            },
+        }
+    }
+
+    pub fn executable_ref(&self) -> Result<TransactionExecutableRef> {
+        match self {
+            TransactionPayload::EntryFunction(entry_function) => {
+                Ok(TransactionExecutableRef::EntryFunction(entry_function))
+            },
+            TransactionPayload::Script(script) => Ok(TransactionExecutableRef::Script(script)),
+            TransactionPayload::Multisig(multisig) => Ok(multisig.as_transaction_executable_ref()),
+            TransactionPayload::Payload(TransactionPayloadInner::V1 { executable, .. }) => {
+                Ok(executable.as_ref())
+            },
+            TransactionPayload::ModuleBundle(_) => {
+                Err(format_err!("ModuleBundle variant is deprecated"))
```

Comment:
this also seems inconsistent with the above

Reply:
Currently, I'm calling the `executable_ref()` function in aptos_vm.rs. When the executable is Err(..), aptos_vm discards the transaction with FEATURE_UNDER_GATING status code.